substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/events/EndChunkNativePeriodicEvents.java

```diff
@@ -66,35 +67,55 @@ private static void emitInitialEnvironmentVariables(StringEntry[] envs) {
             JfrNativeEventWriterData data = StackValue.get(JfrNativeEventWriterData.class);
             JfrNativeEventWriterDataAccess.initializeThreadLocalNativeBuffer(data);
-
+            boolean isLarge = SubstrateJVM.get().isLarge(JfrEvent.InitialEnvironmentVariable);
             for (StringEntry env : envs) {
-                JfrNativeEventWriter.beginEventWrite(data, false);
-                JfrNativeEventWriter.putLong(data, JfrEvent.InitialEnvironmentVariable.getId());
-                JfrNativeEventWriter.putLong(data, JfrTicks.elapsedTicks());
-                JfrNativeEventWriter.putString(data, env.key);
-                JfrNativeEventWriter.putString(data, env.value);
-                JfrNativeEventWriter.endEventWrite(data, false);
+                if (!emitInitialEnvironmentVariable(data, env, isLarge) && !isLarge) {
+                    if (emitInitialEnvironmentVariable(data, env, true)) {
+                        isLarge = true;
+                        SubstrateJVM.get().setLarge(JfrEvent.InitialEnvironmentVariable.getId(), true);
+                    }
+                }
             }
         }
     }
 
+    @Uninterruptible(reason = "Accesses a JFR buffer.")
+    private static boolean emitInitialEnvironmentVariable(JfrNativeEventWriterData data, StringEntry env, boolean isLarge) {
+        JfrNativeEventWriter.beginEventWrite(data, isLarge);
+        JfrNativeEventWriter.putLong(data, JfrEvent.InitialEnvironmentVariable.getId());
+        JfrNativeEventWriter.putLong(data, JfrTicks.elapsedTicks());
+        JfrNativeEventWriter.putString(data, env.key);
+        JfrNativeEventWriter.putString(data, env.value);
+        return JfrNativeEventWriter.endEventWrite(data, isLarge).aboveThan(0);
+    }
+
     @Uninterruptible(reason = "Accesses a JFR buffer.")
     private static void emitInitialSystemProperties(StringEntry[] systemProperties) {
         if (SubstrateJVM.isRecording() && SubstrateJVM.get().isEnabled(JfrEvent.InitialSystemProperty)) {
             JfrNativeEventWriterData data = StackValue.get(JfrNativeEventWriterData.class);
             JfrNativeEventWriterDataAccess.initializeThreadLocalNativeBuffer(data);
-
+            boolean isLarge = SubstrateJVM.get().isLarge(JfrEvent.InitialSystemProperty);
             for (StringEntry systemProperty : systemProperties) {
-                JfrNativeEventWriter.beginEventWrite(data, false);
-                JfrNativeEventWriter.putLong(data, JfrEvent.InitialSystemProperty.getId());
-                JfrNativeEventWriter.putLong(data, JfrTicks.elapsedTicks());
-                JfrNativeEventWriter.putString(data, systemProperty.key);
-                JfrNativeEventWriter.putString(data, systemProperty.value);
-                JfrNativeEventWriter.endEventWrite(data, false);
+                if (!emitInitialSystemProperty(data, systemProperty, isLarge) && !isLarge) {
```

Comment:
I think it is important to document this retry pattern and why it actually works (e.g., in the class-level JavaDoc of `JfrNativeEventWriter`): if the small event got cancelled (e.g., because the flushing fails), then the `JfrNativeEventWriterData` remains in the cancelled state and it is therefore guaranteed that retry fails as well. So, the retry can really only succeed if we really ran out of space while emitting the small event.